### PR TITLE
Ram home card

### PR DESCRIPTION
**Pull request by ramzd18**
This is the finished home card component. It uses fixed values right now for the gym image and data, but when the backend is running it can be changed to incorporate that data. It takes a gym class as a parameter and uses its data to populate the name fields and to also update if a gym is starred by a user.

### CODE REVIEW

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import com.cornellappdev.uplift.R
+import com.cornellappdev.uplift.models.Gym
+import com.cornellappdev.uplift.util.montserratFamily
+
+
+@Composable
+    fun homeCard(gymexample:Gym ) {
+    var paintres = remember  {mutableStateOf(if(gymexample.isFavorite()) R.drawable.ic_star_filled else R.drawable.ic_star) }
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: There are some spacing / formatting errors here--for the sake of keeping our codebase clean, please run the Android Studio code formatter on every file you change. The keyboard shortcut for this is `CTRL + ALT + L` on Windows and `⌘⌥L` on Mac (whatever that means...). That should automatically clean up little spacing errors like this.

(It's a good idea to get in the habit of routinely using this shortcut just to make your code nicer to read while developing, too!)

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                            )
+
+                        }
+                    }
+                }
+            }
+            /**
+             * Changes Icon image to either be filled star or regular star depending on if the user has liked
+             * that particular gym.
+              */
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: I like the comments! But no need for a javadoc comment here--you can just change this to a multiline `//` comment.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import androidx.compose.ui.Alignment.Companion.TopEnd
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import com.cornellappdev.uplift.R
+import com.cornellappdev.uplift.models.Gym
+import com.cornellappdev.uplift.util.montserratFamily
+
+
+@Composable
+    fun homeCard(gymexample:Gym ) {
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: As being enforced by @Corfish123 this semester, please add a javadoc specification comment to any `@Composable` functions you make.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt:1]: Tiny, tiny nitpick--please make `Home` in the filepath lowercase for convention purposes!

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt (hunk 3 above)
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: Also tiny nitpick--please capitalize `homeCard` for convention purposes.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    var paintres = remember  {mutableStateOf(if(gymexample.isFavorite()) R.drawable.ic_star_filled else R.drawable.ic_star) }
+        Box(modifier= Modifier
+            .padding(5.dp)
+            .fillMaxWidth()
+            .fillMaxHeight()
+            .aspectRatio(2F / 1F)){
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: These modifiers in the overarching `Box` cause a problem--since you are telling the `Box` to fill max height BEFORE you set the aspect ratio, the `Box` ends up filling the maximum height of whatever container you put it in, even though you specified the aspect ratio. **Order of modifiers does matter!** I just tested this on a Column with 3 home cards, and they indeed take up way too much height:




To fix this, you can either delete the `fillMaxHeight()` modifier altogether, or put it after the `.aspectRatio()` modifier. You don't need the `fillMaxHeight()` necessarily because the `Box` is already filled to max width and fixed to an aspect ratio, meaning the height should (without a `fillMaxHeight` call) be fixed in height.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            Card(
+                modifier = Modifier
+            ) {
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: Since this `modifier` is the default `Modifier`, please simply delete it as well as the parentheses for code cleanup!

Also, more importantly, it doesn't seem the corner radius on this card follows the Figma designs. Jason's Figma says this should be a corner radius of 12.dp, but yours uses the default 4.dp. To change this, please use the `shape` attribute and `RoundedCornerShape` type (with which you can specify a specific corner radius).

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                Column() {
+                    Column(modifier = Modifier.weight(3F)) {
+                        Image(
+                            modifier = Modifier.fillMaxWidth(),
+                            painter = painterResource(id=R.drawable.court_lines),
+                            contentDescription = "Gym Image",
+                            contentScale = ContentScale.FillWidth
+                        )
+
+                    }
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: Note how the `Image` is the ONLY child in the `Column` here. As such, the column is actually completely unnecessary--you can delete the column and simply move the `.weight()` modifier onto the `Image`. Please do so for code concision.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            .padding(5.dp)
+            .fillMaxWidth()
+            .fillMaxHeight()
+            .aspectRatio(2F / 1F)){
+            Card(
+                modifier = Modifier
+            ) {
+                Column() {
+                    Column(modifier = Modifier.weight(3F)) {
+                        Image(
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: We need to load our images off URLs--as such, we need to use some other component than `Image` here. I chose to use `AsyncImage` to load images from a gym's URL at `gym.imageUrl`--see `GymDetailScreen.kt` for an example, and change this over to an `AsyncImage`.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+                                color = Color.Green,
+                                fontFamily = montserratFamily
+
+                                )
+                            Spacer(modifier = Modifier.width(10.dp))
+                            Text(
+                                text = "Closes at 6 pm",
+                                fontSize = 12.sp,
+                                fontFamily = montserratFamily
+                                )
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: I believe I have provided the sufficient functionality for you to make these dynamically update correctly to a `Gym`'s data. See my function `isCurrentlyOpen()` in `Functions.kt` for determining open/closed status, and change this to update to if a Gym is open or closed at the current time (with correct colors, too). To get a gym's closing time, you can get the last `TimeInterval` appearing in today's `TimeInterval` list. You can get the index of today by doing:

`val day = ((Calendar.getInstance().get(Calendar.DAY_OF_WEEK) - 2) + 7) % 7`

(I should have made a helper function for getting this index, now that I think of it...)

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                                fontSize = 12.sp,
+                                color = Color.Red,
+                                fontFamily = montserratFamily
+                                )
+                            Spacer(modifier = Modifier.width(10.dp))
+                            Text(
+                                text = " 120/140",
+                                fontSize = 12.sp,
+                                fontFamily = montserratFamily
+                            )
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: We will want to make the gym full-ness status dynamically update at some point, of course. I'll take care of this once we get to networking, so don't worry about it for this PR... but let's keep it in mind for the codebase in the future! (Maybe add a TODO here?)

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+                    }
+                }
+            }
+            /**
+             * Changes Icon image to either be filled star or regular star depending on if the user has liked
+             * that particular gym.
+              */
+            IconButton(
+                onClick = {
+                    gymexample.toggleFavorite() },
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: This button seems to have some issues:

1. Its padding is not quite correct according to the Figma.
2. Its colors are off--it's just purely black right now, even when filled. It should be white outlined with a yellow fill.
3. It doesn't seem to actually update when pressed... this is the most confusing issue, and I'll try my best to explain: When you declare `paintres` above, you declare it as a remembered `MutableState` with a default value based on `gymexample.isFavorite()`. As this is a remembered `MutableState`, the value of `gymexample.isFavorite()` is read once, and only once, and therefore never causes recomposition. Then, the value of the state `paintres` never changes to cause recomposition. 

We want to read directly and recompose based off `gymexample.isFavorite()`! You can change this by directly referencing `gymexample.isFavorite()` in your code instead of putting it as the default value of something:

Ex: `var paintres = if (gymexample.isFavorite()) R.drawable.ic_star_filled else R.drawable.ic_star`
(and then, of course, changing `paintres.value` to just `paintres`.)

File: app/src/main/res/values/colors.xml [hunk 11]
```diff
@@ -7,4 +7,8 @@
     <color name="teal_700">#FF018786</color>
     <color name="black">#FF000000</color>
     <color name="white">#FFFFFFFF</color>
+    <color name="light_blue_400">#FF29B6F6</color>
+    <color name="light_blue_600">#FF039BE5</color>
+    <color name="gray_400">#FFBDBDBD</color>
+    <color name="gray_600">#FF757575</color>
```
Comment (thisjustin123) [inline app/src/main/res/values/colors.xml:13]: Are you using these colors anywhere? If not, we can delete them--I've elected to just store colors as constants in `Colors.kt`. Makes life easier, lol

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+                    Column(modifier = Modifier.weight(2F)) {
+                        Row(
+                            modifier = Modifier.padding(5.dp)
+                        ) {
+                            Text(
+                                text =gymexample.name,
+                                fontSize = 20.sp,
+                                fontFamily = montserratFamily
+
+                                )
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: When comparing your implementation to the Figma designs, I see that there are a few coloring, text weight, and font size discrepancies. Please double check text weights and text colors for all these texts. Refer to the fields `fontWeight` and `color` to change those values.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                        Row(
+                            modifier = Modifier.padding(
+                                start = 5.dp,
+                                end = 5.dp
+                            )
+                        ) {
+                            Text(
+                                text = "Cramped",
+                                fontSize = 12.sp,
+                                color = Color.Red,
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: I noticed you used a red color instead of the orange designated on the Figma. Of course, this isn't that big of a deal right now, since we don't need to implement gym capacity right now... but for now, we should probably still import the orange color. Please do so in `Colors.kt` by entering a color as follows:

`val COLOR_NAME = Color(0xFF[HEX CODE HERE])`

**Review by thisjustin123 — CHANGES_REQUESTED** [review]
Overall, a good first PR! However, please address all my comments and keep committing to this branch until they are all satisfied. This component is a **crucial** component in our app, so let's make it as polished as possible! :-)

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt (hunk 3 above)
Comment (Corfish123) [reply app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: damn justin got to this before me :'((((((

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+                                painter = painterResource(id = R.drawable.ic_dumbbell),
+                                contentDescription = "Dumbell",
+                            )
+                            Image(
+                                painter = painterResource(id = R.drawable.ic_bowling_pins),
+                                contentDescription = "Bowling Pins",
+                            )
+
+                        }
+                        Row(modifier = Modifier.padding(start = 5.dp, end = 5.dp)) {
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: You don't need to put padding at the start and end. You can just use `horizontal = 5.dp`

File: app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+                            Text(
+                                text = "Closes at 6 pm",
+                                fontSize = 12.sp,
+                                fontFamily = montserratFamily
+                                )
+
+                        }
+                        Row(
+                            modifier = Modifier.padding(
+                                start = 5.dp,
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/Home/HomeCard.kt]: Same thing, use the horizontal key word for the padding

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+}
+fun determineOpen(listTime: List<List<TimeInterval>?>): Boolean {
+    for (i in 0..listTime.size - 1) {
+        if (isCurrentlyOpen(listTime.get(i)!!)) {
+            return true;
+        }
+    }
+    return false;
+}
+
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt]: Ooh, I think there's a bit of misunderstanding with how the nested list of time intervals work. Here's what I wrote for the spec of gym's `hours` field:

/**
     * A list of exactly 7 lists of time intervals. Each list of time intervals corresponds to a particular
     * day (index 0=Monday, ..., 6=Sunday), and the times in said list indicates the hours of
     * this gym.
     *
     * Example: ((7:00 AM - 8:30 AM, 10:00AM - 10:45PM), ...) indicates that on Monday, this
     * gym is open from 7 to 8:30 AM, then closes, and then is open from 10 AM to 10:45 PM.
     * (These are real gym hours for Teagle Down!)
     *
     * If an index of this list is null, that indicates the gym is closed on that day.
     */
    val hours: List<List<TimeInterval>?>

In other words, you only want to check the list corresponding to today's day of the week. Your `day` local variable is currently getting the correct index--all you have to do now is check for today's day of the week. Right now, you're checking for EVERY day of the week, which could cause problems if other days have different hours. Ultimately, `isCurrentlyOpen(gym.hours(day))` should evaluate to the correct boolean as to if the gym is open or not. 

(I should make this into a helper function in the future...)

Comment (thisjustin123) [reply app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt]: Oh, and since `isCurrentlyOpen(gym.hours(day))` is a relatively short line, I recommend deleting this `determineOpen()` helper function to improve code concision.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+                        model = gymexample.imageUrl,
+                        modifier = Modifier.fillMaxWidth(),
+                        contentDescription = "",
+                        contentScale = ContentScale.FillBounds
+                    )
+                    Image(
+                        painterResource(id = if (gymexample.isFavorite()) R.drawable.ic_star_filled else R.drawable.ic_star),
+                        contentDescription = "Star Icon",
+                        modifier = Modifier
+                            // .size(24.dp)
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt]: delete :-)

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+        modifier = Modifier
+            .padding(horizontal = 16.dp, vertical = 12.dp)
+            .fillMaxWidth()
+            .aspectRatio(2F / 1F)
+    )
+    {
+        Card(
+            shape = RoundedCornerShape(12.dp)
+        ) {
+            Column() {
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt]: Tiny, tiny little nitpick while I'm at it: whenever you have a `@Composable` with empty arguments like this, please try to delete the parentheses as a tiny code concision improvement. I see this a few times throughout this function, so please delete all these you can find.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    Box(
+        modifier = Modifier
+            .padding(horizontal = 16.dp, vertical = 12.dp)
+            .fillMaxWidth()
+            .aspectRatio(2F / 1F)
+    )
+    {
+        Card(
+            shape = RoundedCornerShape(12.dp)
+        ) {
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt]: One weird edge case we need to account for is when the user has Dark Mode on. This causes components to automatically color dark when not given a specified color. Your card on dark mode thus looks like:

This technically looks pretty okay, but until (if ever) design makes dark mode designs, we should stick to light mode strictly. As such, please specify a the direct white color for this card.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+                                indication = null
+                            ) {
+                                gymexample.toggleFavorite()
+                            }
+
+                    )
+                }
+                Column(modifier = Modifier.weight(2F)) {
+                    Row()
+                    {
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt]: Your title area padding is still a bit off. See the following screenshot:

The title, Toni Morrison Gym, is a bit too close to the top. Padding the top of this column should do the trick. (I checked the padding on the Figma, it should be `8.dp`!)

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+                        )
+                        Spacer(modifier = Modifier.width(12.dp))
+                        Spacer(Modifier.height(4.dp))
+                    }
+                    Row() {
+                        Spacer(Modifier.width(12.dp))
+                        if (determineOpen(gymexample.hours)) Text(
+                            text = "Open",
+                            fontSize = 12.sp,
+                            color = Color.Green,
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt]: Tiny nitpick--the design uses a special green instead of this particular green. In the codebase, you can reference this green with `ACCENT_OPEN` from `Colors.kt`.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+                            text = "Cramped",
+                            fontSize = 12.sp,
+                            fontFamily = montserratFamily,
+                            fontWeight = FontWeight(500),
+                            color = colorResource(id = R.color.orange),
+                            modifier = Modifier.padding(top = 2.dp),
+                        )
+                        Spacer(modifier = Modifier.width(8.dp))
+                        Text(
+                            text = " 120/140",
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt:147]: Equally tiny nitpick--the color of this text should be `GRAY03` as specified in the Figma. This also follows for the miles text and "closes at" text.

Oh, also the little bowling icons and dumbbell icons should be `GRAY04`. Please change all these over!

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+        Card(
+            shape = RoundedCornerShape(12.dp)
+        ) {
+            Column() {
+                Box(modifier = Modifier.weight(3F)) {
+                    AsyncImage(
+                        model = gymexample.imageUrl,
+                        modifier = Modifier.fillMaxWidth(),
+                        contentDescription = "",
+                        contentScale = ContentScale.FillBounds
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt]: You probably saw in my screenshots that the image of the gym gets awfully stretched with this content scale. I *think* that we should use `ContentScale.Crop` instead here--this will make the image preserve aspect ratio and simply cut off whatever doesn't fit within the bounds after filling.

**Review by thisjustin123 — CHANGES_REQUESTED** [review]
Really nice work addressing all my comments! I left a few more, most of which are quite small, few-line fixes. Take a look, and submit a final review. Again, great work!

File: app/src/main/java/com/cornellappdev/uplift/MainActivity.kt [hunk 24]
```diff
@@ -14,10 +15,9 @@ class MainActivity : ComponentActivity() {
         val gymDetailViewModel = GymDetailViewModel()
 
         gymDetailViewModel.selectGym(testMorrison)
-
         setContent {
             UpliftTheme {
-                MainNavigationWrapper(gymDetailViewModel = gymDetailViewModel)
+               MainNavigationWrapper(gymDetailViewModel = gymDetailViewModel)
             }
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/MainActivity.kt]: Please run a linter on the codebase before every commit! You can quickly do so by right clicking on our origin package:

image

File: app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt [hunk 25]
```diff
@@ -142,7 +146,8 @@ fun HomeCard(gymexample: Gym) {
                             fontSize = 12.sp,
                             fontFamily = montserratFamily,
                             fontWeight = FontWeight(500),
-                            modifier = Modifier.padding(top = 2.dp)
+                            modifier = Modifier.padding(top = 2.dp),
+                            color = GRAY03
```
Comment (thisjustin123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/home/HomeCard.kt:152]: Please make the `"1.2mi"` text and `"Closes at..."` text `GRAY03` as well!

**Review by thisjustin123 — CHANGES_REQUESTED** [review]
Two tiny, tiny changes, then you're good to go!

**Review by thisjustin123 — APPROVED** [review]
LGTM! 🚀🚀🚀
